applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/eigenfrequency_response_function_utility.h

```diff
@@ -64,327 +64,434 @@ namespace Kratos
+                    aux_vector = Vector(0);;
+                }
+
+                node_i.Y0() -= mDelta;
+                // End derivative of response w.r.t. y-coord --------------------
+
+                // Reset pertubed RHS and mass matrix
+                perturbed_LHS = Matrix(0,0);
+                perturbed_mass_matrix= Matrix(0,0);
+
+                // Derivative of response w.r.t. z-coord ------------------------
+                node_i.Z0() += mDelta;
+
+                elem_i.CalculateMassMatrix(perturbed_mass_matrix, CurrentProcessInfo);
+                perturbed_mass_matrix = (perturbed_mass_matrix - mass_matrix_org) / mDelta;
+
+                elem_i.CalculateLocalSystem(perturbed_LHS, dummy ,CurrentProcessInfo);
+                perturbed_LHS = (perturbed_LHS - LHS_org) / mDelta;
+
+                // Loop over eigenvalues
+                for(std::size_t i = 0; i < num_of_traced_eigenfrequencies; i++)
+                {
+                    aux_matrix = perturbed_LHS;
+                    aux_matrix -= (perturbed_mass_matrix * traced_eigenvalues[i]);
+                    aux_vector = prod(aux_matrix , eigenvectors_of_element[i]);
+                    gradient_contribution[2] += gradient_prefactors[i] * inner_prod(eigenvectors_of_element[i] , aux_vector) * mWeightingFactors[i];
+
+                    aux_matrix = Matrix(0,0);
+                    aux_vector = Vector(0);
+                }
+
+                node_i.Z0() -= mDelta;
+                // End derivative of response w.r.t. z-coord --------------------
 
-				// Derivative of response w.r.t. x-coord ------------------------
-				node_i.X0() += mDelta;
+                // Assemble sensitivity to node
+                noalias(node_i.FastGetSolutionStepValue(SHAPE_SENSITIVITY)) += gradient_contribution;
 
-				elem_i.CalculateMassMatrix(perturbed_mass_matrix, CurrentProcessInfo);
-				perturbed_mass_matrix = (perturbed_mass_matrix - mass_matrix_org) / mDelta;
-				perturbed_mass_matrix *= eigenvalue;
+            }// End loop over nodes of element
 
-				elem_i.CalculateLocalSystem(perturbed_LHS, dummy ,CurrentProcessInfo);
-				perturbed_LHS = (perturbed_LHS - LHS_org) / mDelta;
+        }// End loop over elements
+    }
 
-				perturbed_LHS -= perturbed_mass_matrix;
+    // --------------------------------------------------------------------------
+    double GetEigenvalue(int eigenfrequency_id)
+    {
+        const int num_of_computed_eigenvalues = (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR]).size();
+        return (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR])[eigenfrequency_id-1];
+    }
 
-				aux = prod(perturbed_LHS , eigenvector_of_element);
-				gradient_contribution[0] = inner_prod(eigenvector_of_element , aux);
+    // --------------------------------------------------------------------------
+    Vector GetEigenvectorOfElement(ModelPart::ElementType& traced_element, int eigenfrequency_id, int size_of_eigenvector)
```

Comment:
Passing the eigen_vector by reference and resizing if needed could improve efficiency